### PR TITLE
expose model globally (to Metric etc)

### PR DESCRIPTION
**Pull request by breznak**
Fixes: #1842 

this PR allows to name a model (in constructor() or in `config['modelParams']['name']` of `description.py`) 
and access it from any part of code via `nupic.utils.GlobalDict.get(name)`
Adds abstract class MetricModelCallback and handling for swarming, etc as suggested.

This would unblock work on : AnomalyMetric, SpeedMetric, some additional works on ANomaly class, ...

----------
TODO: 
- [x] CallbackMetric that has access to its Model
- [x] unittest
- [x] choose proper parent class for CallbackMetric and validate the submetric computes something
- [x] simplify some existing Metrics by using sub-metric class

### CODE REVIEW

**Comment by breznak** [comment]
@scottpurdy @oxtopus please review. I try to keep this PR small, it's blocking my further goals now.

**Comment by breznak** [comment]
:grin: what a bad luck! one in 10.000 and I hit it 
https://travis-ci.org/numenta/nupic/jobs/54717668#L4494 
...fixing...

File: nupic/frameworks/opf/model.py [hunk 1]
```diff
@@ -40,17 +41,21 @@ class Model(object):
...
+    @param name (opt) unique ID of the model, if none, random will be generated
     """
     self._numPredictions = 0
     self.__inferenceType =  inferenceType
     self.__learningEnabled = True
     self.__inferenceEnabled = True
     self.__inferenceArgs = {}
+    self._name = name
+    if name is None:
+      self._name = random.randint(0,10000)
```
Comment (oxtopus) [inline nupic/frameworks/opf/model.py]: Maybe do something like `"".join(random.choice(string.ascii_letters) for _ in range(16))` to give it an actual string w/ less chance of overlap?

File: nupic/frameworks/opf/model.py [hunk 2]
```diff
@@ -126,6 +135,34 @@ def _getLogger(self):
...
+    """
+    global globalModelsStorage
+
+    newModel = self
+    logger = logging.getLogger() # self._getLogger() not working because of subclass init
+
+    # generate unique name
+    if newModel._name is None:
+      usedNames = globalModelsStorage.keys()
+      id = random.randint(0,1000)
```
Comment (oxtopus) [inline nupic/frameworks/opf/model.py]: `id` is builtin (https://docs.python.org/2/library/functions.html#id).  For that matter, you could just assign name to some function of `id()`, which will be unique anyway.  For example `"model-%d" % id(newModel)`

**Comment by breznak** [comment]
Thank you @oxtopus ! That's an elegant solution.

**Comment by scottpurdy** [comment]
@breznak - I understand the motivation for this but this isn't the right implementation IMO. The CLAModel has no place keeping track of how many times it has been instantiated, etc. Let's discuss how this stuff should be organized to allow you to do what you are trying to do in #1842

**Comment by breznak** [comment]
@scottpurdy does this sound better? 
Also, `nupic.utils.GlobalDict` might be reused for #1848

**Comment by breznak** [comment]
@scottpurdy @oxtopus 
Thanks for your suggestion, it really simplifies things! Using this example I got it working for anomaly metric (this is a prerequisite for that)
* model<-->metric access is manually handled via `GlobalDict.get()/set()`
* changes to handle models during swarming are in `ModelRunner` 

Please review :wink:

**Comment by oxtopus** [comment]
I think this will work provided swarming happens on a single node in which `GlobalDict` is accessible and consistent, but I'm not sure what will happen when we need to parallelize swarms across multiple nodes.

**Comment by breznak** [comment]
> but I'm not sure what will happen when we need to parallelize swarms across multiple nodes. 

Thanks @oxtopus ! I'm not sure how "Nodes" work, but unless they use 1 python prompt, it would not work. But it works for "personal"-swarming; 
* we could make it work eg by using some sync over the DB
* for simplicity, I'd suggest "multi node parallel-swarming" is just unsupported at the time

What do you think?

**Comment by breznak** [comment]
ping?

**Comment by oxtopus** [comment]
@scottpurdy Do you mind taking a look?  @breznak, I think a merge conflict will need to be resolved.

**Comment by breznak** [comment]
guys, can someone take a look please?

**Comment by breznak** [comment]
@rhyolight @scottpurdy guys, can you review here and #1934 please?

File: nupic/swarming/ModelRunner.py [hunk 3]
```diff
@@ -242,24 +244,40 @@ def run(self):
...
+
+    # special handling for global model sharing (for anomaly metrics,...)
+    # model.name needs to be overriden for swarming by modelID which is unique
+    # this needs to be called before model is created
+    swarmingName = id(self._modelID)
+    modelDescription['modelParams']['name']=swarmingName
+
     self._model = ModelFactory.create(modelDescription)
+    assert GlobalDict.get(swarmingName) is not None
+    print >>sys.stderr,"models=",str(GlobalDict())
```
Comment (rhyolight) [inline nupic/swarming/ModelRunner.py]: Shouldn't this use `self._logger`?

Comment (breznak) [inline nupic/swarming/ModelRunner.py]: I will retest, but I think not. Actually printing to output is somehow
messy when combined with the swarming, standard print is ignored, I copied
this from other use in the code..

**Comment by breznak** [comment]
@rhyolight @scottpurdy ping? I need this to get the anomaly metric running, which I think will be generally useful.

**Comment by breznak** [comment]
@rhyolight this PR is getting dusty, please :walking:

**Comment by david-ragazzi** [comment]
ping [2]

**Comment by rhyolight** [comment]
Yeah, I know. Scott is out of the office.

**Comment by scottpurdy** [comment]
See the issue for relevant discussion:
https://github.com/numenta/nupic/issues/1842#issuecomment-88292787

**Comment by rhyolight** [comment]
@scottpurdy @breznak Wait, I'm confused. Is this PR obsolete due to discussion in #1842? Or are we expecting Marek to continue work? He seems to be waiting for a review of this PR as it is. If there is a miscommunication, we need to work it out so we're not waiting on each other a not knowing it.

**Comment by scottpurdy** [comment]
My opinion on this PR is captured in the comment above:
https://github.com/numenta/nupic/pull/1929#issuecomment-82489644

Austin proposed an alternative:
https://github.com/numenta/nupic/issues/1842#issuecomment-88292787

I thought Austin's proposal was a reasonable short-term solution that doesn't add unnecessary fields to the model. The long-term solution is the generalized Hypersearch library in combination with using the Network API directly (rather than CLAModel).

I'm not sure what @breznak thinks about the options though.

**Comment by rhyolight** [comment]
@scottpurdy Thanks for the clarification. @breznak It seems like you are waiting on use for a PR review, but we are expecting you go implement some changes as discussed by @scottpurdy and @oxtopus above. Please continue that conversation on #1842 and we'll come back to this PR after.

**Comment by breznak** [comment]
guys, please look at the code..it's faster than discussing here :wink:
yes, I've reacted to the Scott & Austins's proposals and this PR : 
- [x] implements "CallbackMetric" as suggested by @oxtopus 
- [x] uses "hack" `GlobalStore` which we agreed is optimal (working & minimally-intrusive short-term solution) as Scott & Austin pointed out

@rhyolight thanks for taking initiative and moving this..

PS: how do I link to the comments here? cheers

File: examples/swarm/anomaly_swarming_hotgym/description.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+"""
+
+from nupic.frameworks.opf.expdescriptionapi import ExperimentDescriptionAPI
+
+from nupic.frameworks.opf.expdescriptionhelpers import (
+  updateConfigFromSubConfig,
+  applyValueGettersToContainer,
+  DeferredDictLookup)
+
+from nupic.frameworks.opf.clamodelcallbacks import *
```
Comment (oxtopus) [inline examples/swarm/anomaly_swarming_hotgym/description.py]: Please use explicit imports

https://github.com/numenta/nupic/wiki/Python-Style-Guide#wildcard-imports

**Comment by rhyolight** [comment]
> PS: how do I link to the comments here? cheers

You can find the anchor links where it says "breznak commented `a day ago`".

File: examples/swarm/anomaly_swarming_hotgym/description.py (hunk 4 above)
Comment (breznak) [inline examples/swarm/anomaly_swarming_hotgym/description.py]: fixed

**Comment by rhyolight** [comment]
This PR is marked as ready for review @scottpurdy @chetan51 @subutai @BoltzmannBrain

**Comment by breznak** [comment]
@rhyolight any progress here?

**Comment by breznak** [comment]
@scottpurdy @chetan51 this should be now clean and ready :)

File: examples/swarm/anomaly_swarming_hotgym/description.py [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+  control['metrics'].append(
+      MetricSpec(field=config['predictedField'], metric='multiStep',
+                 inferenceElement='multiStepBestPredictions',
+                 params={'errorMetric': 'altMAPE', 'window': 1000, 'steps': steps}))
+  control['metrics'].append(
+      MetricSpec(field=config['predictedField'], metric='trivial',
+                 inferenceElement='prediction',
+                 params={'errorMetric': 'altMAPE', 'window': 1000, 'steps': steps}))
+
+################################################################################
```
Comment (david-ragazzi) [inline examples/swarm/anomaly_swarming_hotgym/description.py]: These lines were removed trough https://github.com/numenta/nupic/pull/2249

File: examples/swarm/anomaly_swarming_hotgym/permutations.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+# Permutation optimization setting: either minimize or maximize metric
+# used by RunPermutations.
+# NOTE: The value is used as a regular expressions by RunPermutations.py's
+#       report generator
+# (generated from minimize = 'prediction:aae:window=1000:field=consumption')
+minimize = "multiStepBestPredictions:multiStep:errorMetric='aae':steps=1:window=1000:field=consumption"
+
+
+
+#############################################################################
```
Comment (david-ragazzi) [inline examples/swarm/anomaly_swarming_hotgym/permutations.py]: Same above about removed smugled lines.

File: nupic/frameworks/opf/metrics.py [hunk 7]
```diff
@@ -480,6 +480,31 @@ def _compute(self):
     return self.aggregateError
 
 ################################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: Same above about removed smugled lines.

File: nupic/frameworks/opf/metrics.py [hunk 8]
```diff
@@ -480,6 +480,31 @@ def _compute(self):
...
+    if self.verbosity > 0:
+      print "groundTruth:\n%s\nPredictions:\n%s\n%s\n" % (groundTruth,
+                                            prediction, self.getMetric())
+    # If missing data,
+    if groundTruth == SENTINEL_VALUE_FOR_MISSING_DATA:
+      return self._subErrorMetrics[0].aggregateError
+    # Our "prediction" is simply what submetric does
+    return self._subErrorMetrics[0].addInstance(groundTruth, prediction, record)
+
+################################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: Same above about removed smugled lines.

File: nupic/frameworks/opf/metrics.py [hunk 9]
```diff
@@ -686,7 +711,7 @@ def getMetric(self):
 
 
 ################################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: Same above about removed smugled lines.

File: nupic/frameworks/opf/metrics.py [hunk 10]
```diff
@@ -709,26 +734,14 @@ def __init__(self, metricSpec):
...
     # Use ground truth from 'steps' steps ago as our most recent ground truth
     lastGT = self._getShiftedGroundTruth(groundTruth)
     if lastGT is None:
       return self._subErrorMetrics[0].aggregateError
-
     mean = self._movingAverage(lastGT)
-
-    return self._subErrorMetrics[0].addInstance(groundTruth, mean, record)
+    return super(MetricMovingMean, self).addInstance(groundTruth, mean, record)
+###############################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: If you think this smugled line is really necessary, then just keep it in the midle betwen the 2 methods:

```
def function1():
  return ....


########################


def function2:
  ...
````

File: nupic/frameworks/opf/metrics.py [hunk 11]
```diff
@@ -955,9 +968,8 @@ def addInstance(self, groundTruth, prediction, record = None):
     result = self._subErrorMetrics[0].addInstance(groundTruth, mode, record)
     return result
 
-
-################################################################################
-class MetricTrivial(AggregateMetric):
+###############################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: Same above about removed smugled lines.

File: nupic/frameworks/opf/metrics.py [hunk 12]
```diff
@@ -968,35 +980,14 @@ def __init__(self, metricSpec):
...
-    # If missing data,
-    if groundTruth == SENTINEL_VALUE_FOR_MISSING_DATA:
-      return self._subErrorMetrics[0].aggregateError
-
-    # Our "prediction" is simply what happened 'steps' steps ago
-    return self._subErrorMetrics[0].addInstance(groundTruth, prediction, record)
-
+    return super(MetricTrivial, self).addInstance(groundTruth, prediction, record)
 
 ################################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: Same above about removed smugled lines.

File: nupic/utils.py [hunk 13]
```diff
@@ -114,3 +115,54 @@ def write(self, proto):
     proto.windowSize = self.windowSize
     proto.slidingWindow = self.slidingWindow
     proto.total = self.total
+                                                
+
+#######################################################################
```
Comment (david-ragazzi) [inline nupic/utils.py]: Same above about removed smugled lines.

**Comment by david-ragazzi** [comment]
@breznak I did some comentaries about those smudged lines that were removed in my last PR. I saw a new one that you added to ease the separation, nothing against, but just move it to the midle of the 2 methods to avoid leave it above `def` statement.

**Comment by breznak** [comment]
@david-ragazzi yop, merged. But this is the file where I think the lines were useful. It has so many classes with long methods..

**Comment by david-ragazzi** [comment]
> @david-ragazzi yop, merged. But this is the file where I think the lines were useful. It has so many classes with long methods.. 

opss.. Well, we could create an issue to define which files should have separator and then a PR to fix this.

**Comment by scottpurdy** [comment]
@breznak - I am sorry for the delay on this issue. We will look into this and follow up by the end of this upcoming week.

**Comment by breznak** [comment]
Thanks @scottpurdy ! I rebased on the newest code, and resolved some prev. merge hiccups hopefully. For a short summary, this PR consists of: 
* `MetricSubmetric` - which is just a helper code allowing to clean up some other metrics.
* `MetricModelCallback` - the main code here, allowing to get in instance of active model in a metric, used to "hack" additional info for the metric.
* `GlobalStore` - a helper, wrapper around a global list. I've read your comment somewhere else and agree - this is up for a discussion. 
 * Pros:
   * some prints/debug around the global list
   * user just needs to know there's `utils.GlobalStore with get()/set()`
   * also all code is wrapped just in one class logic - when we want to get rid of this functionality, we'll just remove all mentions of GlobalStore, no problems finding global variables with custom names.
 * Cons: 
   * additional cruft code - agreed the added functionality is not something that cannot be really handeled by just an array. 

So yeah, GlobalStore has questionable and I will easily replace it with just some global variable. I just find the functionality less intrusive when packed into 1 simple code.

**Comment by scottpurdy** [comment]
@breznak - where can we centralize discussions about these things? @chetan51 thought through how to support metrics that use the time. And we should just extend NuPIC to support other metric implementations and you can do the rest with that. We can avoid including experimental metrics in the main repo that way.

**Comment by breznak** [comment]
Here? https://github.com/numenta/nupic/issues/2074#issuecomment-117374635

**Comment by breznak** [comment]
Closing as discussion in https://github.com/numenta/nupic/issues/1711#issuecomment-119359824 
now suggests we prefer to add necessary fields to ModelResult rather than hacks to CLAModel instance. 
TODO: pick MetricSubmetric and other doc/cleanup commits from this PR.

**Pull request closed without merging**